lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+  /// \tparam RegisterCount the number of registers for the current ABI, should
+  ///         be auto-magically deduced from the \ref Allowed.
+  ///
+  /// \return
   template<typename RegisterType, size_t RegisterCount, bool DryRun = false>
   static std::optional<model::QualifiedType>
-  convertReturnValue(const SortedVector<RegisterType> &UsedRegisters,
-                     const RegisterArray<RegisterCount> &AllowedRegisters,
-                     const model::Register::Values PointerToCopyLocation,
-                     model::Binary &TheBinary) {
-    if (UsedRegisters.size() == 0) {
-      auto Void = TheBinary.getPrimitiveType(model::PrimitiveTypeKind::Void, 0);
-      return model::QualifiedType{ Void, {} };
+  tryConvertingReturnValue(const SortedVector<RegisterType> &Used,
+                           const RegisterArray<RegisterCount> &Allowed,
+                           const model::Register::Values PointerToCopyLocation,
+                           const model::Binary &Binary,
+                           TypeVector &TypeDependencies) {
+    if (Used.size() == 0) {
+      // Short-circuit the functions that don't return anything (`void`).
+      return ensurePrimitiveTypeIsAvailable(model::PrimitiveTypeKind::Void,
+                                            0,
+                                            Binary,
+                                            TypeDependencies);
+    }
+
+    if (Allowed.empty()) {
+      // A return value is expected on an ABI that doesn't dedicate any
+      // registers to be used for those.
+      return std::nullopt;
     }
 
-    if (UsedRegisters.size() == 1) {
-      if (UsedRegisters.begin()->Location == PointerToCopyLocation) {
-        if constexpr (DryRun)
-          return model::QualifiedType{};
-        else
-          return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                  PointerToCopyLocation,
-                                  TheBinary);
+    // Special case where the only used register is the pointer-to-copy location
+    if (Used.size() == 1 && Used.begin()->Location == PointerToCopyLocation) {
+      if (Used.begin()->Type.isPointer()) {
+        // If the type is already a pointer, preserve it.
+        return Used.begin()->Type;
       } else {
-        if constexpr (RegisterCount == 0)
-          return std::nullopt;
-        if (AllowedRegisters.front() == UsedRegisters.begin()->Location) {
-          if constexpr (DryRun)
-            return model::QualifiedType{};
-          else
-            return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                    UsedRegisters.begin()->Location,
-                                    TheBinary);
-        } else {
-          return std::nullopt;
-        }
+        // Otherwise return a `void *` pointer as the return value type.
+        static constexpr auto VoidKind = model::PrimitiveTypeKind::Void;
+        auto Result = ensurePrimitiveTypeIsAvailable(VoidKind,
```

Comment:
I'd prefer to with `generic8_t *`. We have plans to use `void *` in a special way, i.e., the return type of `malloc`. It'll basically indicate not to propagate the type around, the function might return a different actual type at every call.

Reply:
Sounds good. But I think I'd prefer `generic32_t *`/`generic64_t *` instead, since these indicate the most common alignment found in the structs/arrays the pointer likely points to.

Reply:
Mmmh, `generic8_t` gives a better idea that we don't know what we're pointing to. Using something else feeds false hopes `:D`

Reply:
Yeah, you're right, it would be cruel to make someone believe that there's more information than there actually is.

Reply:
Despite the name, rev.ng is a analyst-cruelty free project.

Reply:
🤞 What's left is for others to preserve us as such too 🤞

Reply:
This is done.